File: frb_example/pure_dart/lib/src/rust/api/comment.dart
// ignore_for_file: invalid_use_of_internal_member

import '../frb_generated.dart';

/// Multiline comments are fine,
/// but they are not preferred in Rust nor in Dart.
/// Newlines are preserved.
Future<void> functionWithCommentsSlashStarStarTwinNormal({dynamic hint}) =>
    RustLib.instance.api
        .functionWithCommentsSlashStarStarTwinNormal(hint: hint);

/// This is first line
/// This is second line
Future<void> functionWithCommentsTripleSlashMultiLineTwinNormal(
        {dynamic hint}) =>
    RustLib.instance.api
        .functionWithCommentsTripleSlashMultiLineTwinNormal(hint: hint);

/// This is single line comment
Future<void> functionWithCommentsTripleSlashSingleLineTwinNormal(
        {dynamic hint}) =>
    RustLib.instance.api
        .functionWithCommentsTripleSlashSingleLineTwinNormal(hint: hint);

/// Comments on structs
class StructWithCommentsTwinNormal {
  /// Documentation on a struct field
  final int fieldWithComments;

  const StructWithCommentsTwinNormal({
    required this.fieldWithComments,
  });

  /// Documentation on an instance method
  Future<void> instanceMethod({dynamic hint}) =>
      RustLib.instance.api.structWithCommentsTwinNormalInstanceMethod(
        that: this,
      );

  /// Documentation on a static method
  static Future<void> staticMethod({dynamic hint}) =>
      RustLib.instance.api.structWithCommentsTwinNormalStaticMethod(hint: hint);
}

File: frb_example/pure_dart/rust/src/api/comment.rs
/// This is single line comment
pub fn function_with_comments_triple_slash_single_line_twin_normal() {}

/// This is first line
/// This is second line
pub fn function_with_comments_triple_slash_multi_line_twin_normal() {}

/**
 Multiline comments are fine,
 but they are not preferred in Rust nor in Dart.
 Newlines are preserved.
*/
pub fn function_with_comments_slash_star_star_twin_normal() {}

/// Comments on simple enums
pub enum SimpleEnumWithCommentsTwinNormal {
    Apple,
    /// Comments on enum variants
    Orange,
}

/// Comments on complex enums
pub enum ComplexEnumWithCommentsTwinNormal {
    Apple,
    Orange {
        /// Comments on enum variant's fields
        a_field: i32,
    },
    Raspberry(
        /// Comments on enum variant's anonymous fields
        i32,
    ),
}

/// Comments on structs
pub struct StructWithCommentsTwinNormal {
    /// Documentation on a struct field
    pub field_with_comments: i32,
}

impl StructWithCommentsTwinNormal {
    /// Documentation on a static method
    pub fn static_method_twin_normal() {}

    /// Documentation on an instance method
    pub fn instance_method_twin_normal(&self) {}
}
